Fb sm test fix for units field

labkey-danield opened the pull request:
#### Rationale
Ran across an issue with the grid component when writing Sample Manager test. If the grid is long, and has been scrolled,  the previous and next buttons can be scrolled out of view. Need to scroll them into view before clicking them.

#### Related Pull Requests
* https://github.com/LabKey/sampleManagement/pull/393

#### Changes
* Added a scrollIntoView in clickPrevious and clickNext.

labkey-danield committed:
Merge branch 'develop' into fb_smTestFixForUnitsField

labkey-danield committed:
Add selector for required field in sample type creation.
Sample Delete warning message in dialog is different when freezer manager is turned on.
Grid bar elements need to be visible to click them.
Fix Workflow test to have consistent samples selected.

labkey-danield committed:
Merge branch 'develop' into fb_smTestFixForUnitsField

labkey-danield committed:
Merge branch 'develop' into fb_smTestFixForUnitsField

labkey-tchad commented on src/org/labkey/test/components/react/Pager.java:
```diff
@@ -63,8 +63,11 @@ public int getPageSize()                // only works on GridPanel
     public Pager clickPrevious()
     {
         if (elementCache().prevButton().isPresent())
-            _pagedComponent.doAndWaitForUpdate(()->
+        {
+            getWrapper().scrollIntoView(elementCache().prevButton().get());
+            _pagedComponent.doAndWaitForUpdate(() ->
                     elementCache().prevButton().get().click());
+        }
```
We're finding the button three times here. We should avoid that.
```suggestion
        elementCache().prevButton().ifPresent(button -> {
            _pagedComponent.doAndWaitForUpdate(()-> {
                getWrapper().scrollIntoView(button);
                button.click();
            });
        });
```

labkey-danield replied:
I was unaware of the ifPresent option.
I have made the update, but it does touch on an issue we discussed in our coding standards meeting. If the previous, or next button, is not there this code will do nothing. Does that run counter to what we had talked about?

labkey-tchad commented on src/org/labkey/test/components/react/Pager.java:
```diff
@@ -83,8 +86,11 @@ public boolean isPreviousEnabled()
     public Pager clickNext()
     {
         if (elementCache().prevButton().isPresent())
-            _pagedComponent.doAndWaitForUpdate(()->
+        {
+            getWrapper().scrollIntoView(elementCache().nextButton().get());
+            _pagedComponent.doAndWaitForUpdate(() ->
                     elementCache().nextButton().get().click());
+        }
```
Find button once.
```suggestion
        elementCache().nextButton().ifPresent(button -> {
            _pagedComponent.doAndWaitForUpdate(()-> {
                getWrapper().scrollIntoView(button);
                button.click();
            });
        });
```

labkey-danield replied:
Implemented the change.

labkey-tchad reviewed (approved):
Looks ok.
The suggestion here is for a pre-existing problem since you're modifying this file already. Perhaps apply similar changes to `isNext/PreviousEnabled`.

labkey-danield committed:
Merge branch 'develop' into fb_smTestFixForUnitsField

labkey-danield committed:
Implement code review feedback.

labkey-chrisj reviewed (approved):
This seems sensible.  Thanks!

labkey-danield committed:
Merge branch 'develop' into fb_smTestFixForUnitsField

labkey-danield committed:
Implement code review feedback.
Moving InventoryMeasurementUnit to the Inventory module.
Moving the units field to the SampleTypeDefinition in testAutomation.
Created interface MetricUnit to allow bridge updates to testAutomation, Inventory and SampleManager modules.

labkey-tchad commented on src/org/labkey/test/params/experiment/MetricUnit.java:
```diff
@@ -0,0 +1,7 @@
+package org.labkey.test.params.experiment;
+
+public interface MetricUnit<T extends MetricUnit<T>>
```
This sort of self-referential type parameter is just a hack we do to allow chaining in some pages and components. It shouldn't be necessary here.
```suggestion
public interface MetricUnit
```

labkey-danield replied:
Removed. I was trying to work around an issue and forgot I had added this.

labkey-tchad commented on src/org/labkey/test/params/experiment/SampleTypeDefinition.java:
```diff
@@ -30,6 +30,9 @@
     // Indicates which parent aliases reference 'exp.dataInputs' instead of 'exp.materialInputs'
     private Set<String> _dataParentAliases = new HashSet<>();
 
+    private MetricUnit _inventoryMetricUnit;
+    private String _labelColor;
```
Maybe add a note indicating that these are used only by Freezer Manager and Sample Manager

labkey-danield replied:
Updated. I limited the comment to the SampleManager module because that is the only module that takes advantage of them.

labkey-tchad commented on src/org/labkey/test/params/experiment/SampleTypeDefinition.java:
```diff
@@ -68,6 +71,28 @@ public SampleTypeDefinition setDescription(String description)
+        _labelColor = color;
+        return this;
+    }
+
```
To make these work with API helpers, add `metricUnit` and `labelColor` to the map produced by `getOptions`:
```
if (getInventoryMetricUnit() != null)
{
    options.put("metricUnit", getInventoryMetricUnit());
}
if (getLabelColor() != null)
{
    options.put("labelColor", getLabelColor());
}
```

labkey-danield replied:
Updated.

labkey-tchad reviewed (approved):
Suggestion to make the new functionality work over API

labkey-danield committed:
Minor clean-up.
Changing names to be more consistent.
Added comments.
Made return type and parameters for getter/setter consistent.
Removed cast operation.
